internal/fwserver/attribute_validation.go

```diff
@@ -101,47 +101,97 @@ func AttributeValidate(ctx context.Context, a fwschema.Attribute, req ValidateAt
 		return
 	}
 
-	// Terraform CLI does not automatically perform certain configuration
-	// checks yet. If it eventually does, this logic should remain at least
-	// until Terraform CLI versions 0.12 through the release containing the
-	// checks are considered end-of-life.
-	// Reference: https://github.com/hashicorp/terraform/issues/30669
-	if a.IsComputed() && !a.IsOptional() && !attributeConfig.IsNull() {
-		resp.Diagnostics.AddAttributeError(
-			req.AttributePath,
-			"Invalid Configuration for Read-Only Attribute",
-			"Cannot set value for this attribute as the provider has marked it as read-only. Remove the configuration line setting the value.\n\n"+
-				"Refer to the provider documentation or contact the provider developers for additional information about configurable and read-only attributes that are supported.",
-		)
-	}
+	// Dynamic values need to perform more logic to check the config value for null/unknown-ness
+	dynamicValuable, ok := attributeConfig.(basetypes.DynamicValuable)
+	if ok {
+		dynamicConfigVal, diags := dynamicValuable.ToDynamicValue(ctx)
+		resp.Diagnostics.Append(diags...)
+		if diags.HasError() {
+			return
+		}
 
-	// Terraform CLI does not automatically perform certain configuration
-	// checks yet. If it eventually does, this logic should remain at least
-	// until Terraform CLI versions 0.12 through the release containing the
-	// checks are considered end-of-life.
-	// Reference: https://github.com/hashicorp/terraform/issues/30669
-	if a.IsRequired() && attributeConfig.IsNull() {
-		resp.Diagnostics.AddAttributeError(
-			req.AttributePath,
-			"Missing Configuration for Required Attribute",
-			fmt.Sprintf("Must set a configuration value for the %s attribute as the provider has marked it as required.\n\n", req.AttributePath.String())+
-				"Refer to the provider documentation or contact the provider developers for additional information about configurable attributes that are required.",
-		)
-	}
+		// Terraform CLI does not automatically perform certain configuration
+		// checks yet. If it eventually does, this logic should remain at least
+		// until Terraform CLI versions 0.12 through the release containing the
+		// checks are considered end-of-life.
+		// Reference: https://github.com/hashicorp/terraform/issues/30669
+		if a.IsComputed() && !a.IsOptional() && !dynamicConfigVal.IsUnderlyingValueNull() {
```

Comment:
I believe what you're likely observing with the corner provider errors are that you also need to check the container (`attributeConfig.IsNull()`) for null/not nullness as well for dynamics. If you refactor to putting this into a variable we can avoid duplicating the logic as well 👍🏻

Reply:
I definitely missed adding any unit tests for these scenarios when I implemented dynamic types, which makes sense because the logic was broken 😆